### DISCUSSION TITLE
Callback for table headers?

### DISCUSSION THREAD

**Discussion by CramBL** [fltk-rs/fltk-rs#1466] [Q&A] [answered]
I have a table that I want to make more intuitive to use, for example adding some kind of sorting icon to the column headers and when you click them the table is sorted by the values of that column, is that possible?

I can already sort, and I can just add buttons that sort by various values, but it is not as nice. Basically, I end up making a button for each header which is a shame and a waste of space.

Can I just register a callback for headers? I've looked through the table crate and it does not seem to be implemented yet or am I missing something?

**Comment by MoAlyousef** [comment, marked as answer]
You should be able to import the TableExt trait and use the callback_context() method:
```rust
use fltk::{
    app, enums,
    prelude::{GroupExt, TableExt, WidgetExt},
    table::TableContext,
    window,
};
use fltk_table::{SmartTable, TableOpts};

fn main() {
    let app = app::App::default().with_scheme(app::Scheme::Gtk);
    let mut wind = window::Window::default().with_size(800, 600);

    // We pass the rows and columns thru the TableOpts field
    let mut table = SmartTable::default()
    .with_size(790, 590)
    .center_of_parent()
    .with_opts(TableOpts {
        rows: 30,
        cols: 30,
        editable: true,
        ..Default::default()
    });

    wind.end();
    wind.show();

    table.set_callback(move |t| {
        if t.callback_context() == TableContext::ColHeader {
            if app::event() == enums::Event::Released {
                dbg!(t.col_header_value(t.callback_col()));
            }
        }
    });


    app.run().unwrap();
}
```

Adding an icon for sorting might be more involved. I do it here for a HoldBrowser:
https://github.com/MoAlyousef/sysinfo-gui/blob/main/src/view/procs.rs

The idea for a Table is to not create column headers, and instead create a pack of group widgets, with an icon radio button inside them. You can use the Table::find_cell to query the coordinates of your groups. 
I haven’t tried doing this with tables though. 

A simpler option is to use a hidden MenuButton and you just call popup() on it when the user right clicks the header.

**Reply by CramBL** [reply]
Thanks a lot! And thanks for the link to a great example, I'm sure I can make it work from here.

Edit: Works perfectly. However, to be highly configurable (sort by multiple columns at a time, and show symbols depending on it) it involves managing state for each column (`Arc<Mutex<ColumnState>>` per Column). Encapsulating it in a `ColumnState` struct made it manageable though.
